src/Avalonia.Visuals/Media/Pen.cs

```diff
@@ -207,13 +208,23 @@ void UpdateSubscription(ref IAffectsRender? field, object? value)
         {
             if ((_invalidated == null || field != value) && field != null)
             {
-                InvalidatedWeakEvent.Unsubscribe(field, this);
+                if (_weakSubscriber != null)
+                    InvalidatedWeakEvent.Unsubscribe(field, _weakSubscriber);
                 field = null;
             }
 
             if (_invalidated != null && field != value && value is IAffectsRender affectsRender)
             {
-                InvalidatedWeakEvent.Subscribe(affectsRender, this);
+                if (_weakSubscriber == null)
```

Comment:
If understand correctly (and I have to guess since you didn’t provide any PR description) this is meant to make weak event handling private? I’m wondering if a subscriber that simply takes a delegate that has no capture (and has a target object) isn’t a better solution. `Event` seems to have a single subscriber anyway, lambda below has to allocate a new delegate array, a new delegate and a closure due to `this` capture.

Reply:
> this is meant to make weak event handling private

Yes, technically implemented interface list is a part of the public API and we shouldn't expose such details for the outside world

>I’m wondering if a subscriber that simply takes a delegate that has no capture (and has a target object) isn’t a better solution

Yes, that's certainly an option since it saves us at least one extra allocation.

Reply:
Do you have time to do this or should I take it over?

Reply:
I've taken over this PR now. Implemented a solution that minimizes extra allocations (and drops one leaky closure allocation that was happening even when weak subscriber was already there).